webserver/src/main/java/io/kestra/webserver/controllers/api/KVController.java

```diff
@@ -0,0 +1,113 @@
+package io.kestra.webserver.controllers.api;
+
+import io.kestra.core.exceptions.ResourceExpiredException;
+import io.kestra.core.serializers.JacksonMapper;
+import io.kestra.core.storages.StorageInterface;
+import io.kestra.core.storages.kv.InternalKVStore;
+import io.kestra.core.storages.kv.KVEntry;
+import io.kestra.core.storages.kv.KVMetadata;
+import io.kestra.core.storages.kv.KVStoreValueWrapper;
+import io.kestra.core.tenant.TenantService;
+import io.micronaut.http.HttpHeaders;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.annotation.*;
+import io.micronaut.scheduling.TaskExecutors;
+import io.micronaut.scheduling.annotation.ExecuteOn;
+import io.micronaut.validation.Validated;
+import io.swagger.v3.oas.annotations.Operation;
+import io.swagger.v3.oas.annotations.Parameter;
+import jakarta.inject.Inject;
+
+import java.io.*;
+import java.net.URISyntaxException;
+import java.time.*;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+@Validated
+@Controller("/api/v1/namespaces/{namespace}/kv")
+public class KVController {
+    @Inject
+    private StorageInterface storageInterface;
+    @Inject
+    private TenantService tenantService;
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get
+    @Operation(tags = {"KV"}, summary = "List all keys for a namespace")
+    public List<KVEntry> list(
+        @Parameter(description = "The namespace id") @PathVariable String namespace
+    ) throws IOException, URISyntaxException {
+        return kvStore(namespace).list();
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "{key}")
+    @Operation(tags = {"KV"}, summary = "Get value for a key")
+    public TypedValue get(
+        @Parameter(description = "The namespace id") @PathVariable String namespace,
+        @Parameter(description = "The key") @PathVariable String key
+    ) throws IOException, URISyntaxException, ResourceExpiredException {
+        Object value = kvStore(namespace).get(key).orElseThrow(() -> new NoSuchElementException("No value found for key '" + key + "' in namespace '" + namespace + "'"));
+        return new TypedValue(KVType.from(value), value);
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Put(uri = "{key}", consumes = {MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON})
+    @Operation(tags = {"KV"}, summary = "Puts a key-value pair in store")
+    public void put(
+        HttpHeaders httpHeaders,
+        @Parameter(description = "The namespace id") @PathVariable String namespace,
+        @Parameter(description = "The key") @PathVariable String key,
+        @Body String value
+    ) throws IOException, URISyntaxException, ResourceExpiredException {
+        String ttl = httpHeaders.get("ttl");
```

Comment:
I would have add a parameter like for everything else as headers are usually for "technical stuff" and here TTL is a property of a KV.

Reply:
I don't really want to include the ttl in the body as it will be a pain to handle all the datatypes (I'll have to get everything as a String, check if that some json value, I'll probably lose some typing if I have a JSON wrapper everytime as in theory I then must have always a MediaType JSON then but then I was basing off that to know if I should convert to object then back to ion.

So a bit of technical constraint but aside of that I honestly don't see a true issue in ttl being in headers. I mean we store it in metadata in storage and I've already seen a lot of soft that are expecting such properties in headers :thinking: